internal/webhook/handlers/tcp_gateway_validate.go

```
@@ -0,0 +1,82 @@
+// Copyright 2022 Clastix Labs
+// SPDX-License-Identifier: Apache-2.0
+
+package handlers
+
+import (
+	"context"
+	"fmt"
+
+	"gomodules.xyz/jsonpatch/v2"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/discovery"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+
+	kamajiv1alpha1 "github.com/clastix/kamaji/api/v1alpha1"
+	"github.com/clastix/kamaji/internal/utilities"
+	"github.com/clastix/kamaji/internal/webhook/utils"
+)
+
+type TenantControlPlaneGatewayValidation struct {
+	Client          client.Client
+	DiscoveryClient discovery.DiscoveryInterface
+}
+
+func (t TenantControlPlaneGatewayValidation) OnCreate(object runtime.Object) AdmissionResponse {
+	return func(ctx context.Context, _ admission.Request) ([]jsonpatch.JsonPatchOperation, error) {
+		tcp := object.(*kamajiv1alpha1.TenantControlPlane)
+
+		if tcp.Spec.ControlPlane.GatewayRoute != nil {
+			// NOTE: Do we actually want to deny here if Gateway API is not available or a warning?
```

From an operation perspective, I think denying can be problematic. I'd rather accept and set a status that indicates that the gateway is not available?

---

I'm not sure about this, if the spec has Gateway stuff but no Gateway API are present isn't it nice to let it fail and then the user can act on it? Instead of having to find out through the status?

---

We don't report errors in Status for Gateway objects; we need to block it here.